common/shamir-secret.c

```diff
@@ -0,0 +1,267 @@
+    }
+
+    if (mpz_cmp(mod, f1) == 0 || mpz_cmp(mod, f2) == 0) {
+        mpz_set_ui(result, 0);
+        return;
+    }
+
+    mpz_set_ui(mask1, 1);
+    mpz_mul_2exp(mask1, mask1, 256);
+    mpz_set(v, f1);
+    mpz_set_ui(z, 0);
+    while (mpz_cmp_ui(f2, 0) != 0) {
+        mpz_set_ui(mask2, mpz_tstbit(f2, 0));
+        if (mpz_cmp_ui(mask2, 0) > 0) {
+            char *msk = (char *)malloc(256 + 1);
+            memset(msk, '1', 256);
+            msk[256] = '\0';
+            mpz_set_str(mask2, msk, 2);
+            free(msk);
+        }
+
+        mpz_t tmp, tmp2;
+        mpz_inits(tmp, tmp2, NULL);
+        mpz_xor(tmp, z, v);
+        mpz_and(tmp, tmp, mask2);
+
+        mpz_sub(tmp2, mask1, mask2);
+        mpz_sub_ui(tmp2, tmp2, 1);
+        mpz_and(tmp2, z, tmp2);
+        mpz_ior(z, tmp, tmp2);
+
+        mpz_mul_2exp(v, v, 1);
+
+        mpz_div_2exp(tmp2, v, 256);
+        mpz_set_ui(mask3, mpz_tstbit(tmp2, 0));
+        if (mpz_cmp_ui(mask3, 0) > 0) {
+            char *msk = (char *)malloc(256 + 1);
+            memset(msk, '1', 256);
+            msk[256] = '\0';
+            mpz_set_str(mask3, msk, 2);
+            free(msk);
+        }
+
+        mpz_xor(tmp, v, mod);
+        mpz_and(tmp, tmp, mask3);
+
+        mpz_sub(tmp2, mask1, mask3);
+        mpz_sub_ui(tmp2, tmp2, 1);
+        mpz_and(tmp2, tmp2, v);
+        mpz_ior(tmp2, tmp, tmp2);
+        mpz_set(v, tmp2);
+
+        mpz_clear(tmp);
+        mpz_clear(tmp2);
+        mpz_div_2exp(f2, f2, 1);
+    }
+    mpz_set(result, z);
+    
+    mpz_clear(f1);
+    mpz_clear(f2);
+    mpz_clear(mask1);
+    mpz_clear(mask2);
+    mpz_clear(mask3);
+    mpz_clear(v);
+    mpz_clear(z);
+}
+
+
+struct point *split(int n, int t, char *secret) {
+    struct point *points = malloc(n * sizeof(struct point));
+    mpz_t sec;
+    mpz_t coefficients[t];
+    gmp_randstate_t state;
+    mpz_t mod;
+    mpz_init_set_str(mod, prime_polynomial_str, 10);
+
+    mpz_init(sec);
+    if (mpz_set_str(sec, secret, 16) != 0) {
+        fprintf(stderr, "Invalid hex string\n");
+        mpz_clear(sec);
+        return NULL;
+    }
+
+    gmp_randinit_default(state);
+    for (int i = 0; i < t - 1; i++) {
+        mpz_init(coefficients[i]);
+        mpz_urandomb(coefficients[i], state, 256);
+    }
+    mpz_init_set(coefficients[t - 1], sec);
+
+    gmp_randclear(state);
+    
+    for (int i = 1; i < n + 1; i++) {
+        mpz_t share, tmp;
+        mpz_init(tmp);
+        mpz_init_set_ui(share, 0);
+        mpz_t idx;
+        mpz_init_set_ui(idx, i);
+
+        for (int j = 0; j < t; j++) {
+            mul_gf2256(tmp, share, idx);
+            mpz_set(share, tmp);
+            mpz_xor(share, share, coefficients[j]);
+        }
+
+        mpz_init_set_ui(points[i-1].x, i);
+        mpz_init_set(points[i-1].y, share);
+
+        mpz_clear(share);
+        mpz_clear(idx);
+        mpz_clear(tmp);
+    }
+    return points;
+}
+
+void combine (struct point *points, int n) {
```

Comment:
I think this should return the secret, right?

So:

const struct secret *shamir_combine(const tal_t *ctx, const struct point *points, size_t num_points);

And what happens if I don't give it enough points for the threshold?